File: services/product-service/config/s3.go
package config

import (
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

var S3Client *s3.S3
var S3BucketName string

func InitS3() {
	region := os.Getenv("AWS_REGION")
	accessKey := os.Getenv("AWS_ACCESS_KEY_ID")
	secretKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
	bucketName := os.Getenv("AWS_S3_BUCKET_NAME")
	endpoint := os.Getenv("AWS_S3_ENDPOINT") // Optional: for MinIO or LocalStack

	fmt.Printf("🔍 S3 Config - Region: %s, Bucket: %s, Endpoint: %s\n", region, bucketName, endpoint)

	if region == "" || accessKey == "" || secretKey == "" || bucketName == "" {
		fmt.Println("⚠️  Missing AWS S3 configuration (AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_S3_BUCKET_NAME)")
		return
	}

	S3BucketName = bucketName

	// Build AWS config with timeout
	awsConfig := &aws.Config{
		Region:      aws.String(region),
		Credentials: credentials.NewStaticCredentials(accessKey, secretKey, ""),
		HTTPClient: &http.Client{
			Timeout: 5 * time.Second, // 5 second timeout to prevent hanging
		},
	}

	// Add endpoint if specified (for MinIO, LocalStack, etc.)
	if endpoint != "" {
		awsConfig.Endpoint = aws.String(endpoint)
		awsConfig.S3ForcePathStyle = aws.Bool(true)       // Required for MinIO and some S3-compatible services
		awsConfig.DisableSSL = aws.Bool(true)             // Disable SSL for local development
		fmt.Printf("🔧 Using custom S3 endpoint: %s (SSL disabled)\n", endpoint)
	}

	sess, err := session.NewSession(awsConfig)
	if err != nil {
		fmt.Printf("❌ Failed to create AWS session: %v\n", err)
		return
	}

	S3Client = s3.New(sess)
	fmt.Println("✅ S3 client initialized")

	// Test connection by listing buckets
	fmt.Println("🧪 Testing S3 connection...")
	buckets, err := S3Client.ListBuckets(&s3.ListBucketsInput{})
	if err != nil {
		fmt.Printf("❌ Failed to list S3 buckets: %v\n", err)
		fmt.Println("💡 Common issues:")
		fmt.Println("   - Check if MinIO/S3 service is running")
		fmt.Println("   - Verify AWS_S3_ENDPOINT URL is correct")
		fmt.Println("   - Ensure credentials are valid")
		return
	}
	
	fmt.Printf("✅ Successfully connected to S3! Found %d bucket(s)\n", len(buckets.Buckets))
	for _, bucket := range buckets.Buckets {
		fmt.Printf("   📦 %s\n", *bucket.Name)
	}
}

File: services/notification-service/config/s3.js
/**
 * AWS S3 Configuration
 * Initialize S3 client for file uploads
 */

import { S3Client } from '@aws-sdk/client-s3';
import dotenv from 'dotenv';

dotenv.config();

let s3Client = null;
let s3BucketName = null;

/**
 * Initialize S3 client with credentials from environment
 */
export const initS3 = () => {
  const region = process.env.AWS_REGION;
  const accessKeyId = process.env.AWS_ACCESS_KEY_ID;
  const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
  const bucketName = process.env.AWS_S3_BUCKET_NAME;
  const endpoint = process.env.AWS_S3_ENDPOINT; // Optional for MinIO/LocalStack

  console.log(`🔍 S3 Config - Region: ${region}, Bucket: ${bucketName}, Endpoint: ${endpoint || 'default'}`);

  if (!region || !accessKeyId || !secretAccessKey || !bucketName) {
    console.warn('⚠️  Missing AWS S3 configuration. Image upload will not be available.');
    console.warn('   Required: AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_S3_BUCKET_NAME');
    return;
  }

  s3BucketName = bucketName;

  const config = {
    region,
    credentials: {
      accessKeyId,
      secretAccessKey,
    },
  };

  // Add custom endpoint if specified (for MinIO, LocalStack, etc.)
  if (endpoint) {
    config.endpoint = endpoint;
    config.forcePathStyle = true; // Required for MinIO and some S3-compatible services
    console.log(`🔧 Using custom S3 endpoint: ${endpoint}`);
  }

  s3Client = new S3Client(config);
  console.log('✅ S3 client initialized successfully');
};

/**
 * Get S3 client instance
 * @returns {S3Client|null}
 */
export const getS3Client = () => s3Client;

/**
 * Get S3 bucket name
 * @returns {string|null}
 */
export const getS3BucketName = () => s3BucketName;

export default { initS3, getS3Client, getS3BucketName };
